index.php

```diff
@@ -439,7 +395,159 @@ function locate_file_for_ebookmaker(string $dirname): string
     return $basename;
 }
 
-function exception_handler($exception): void
+function validate_html($basename)
+{
+    ##### Figure out HTML version if needed, and generate output for user:
+    $not_html5 = 0;
+    if (str_ends_with($basename, ".html") || str_ends_with($basename, ".htm")) {
+        # Did we get HTML5? Look at the first line:
+        $tmpin = fopen($basename, "r");
+        if ($tmpin == false) {
+            throw new RuntimeException("Error opening file $basename for validation.");
+        }
+        $line = fgets($tmpin);
+        $regex = '/<\!DOCTYPE html/i';
+        if (! preg_match($regex, $line)) {
+            $not_html5 = 1;
+        }
+        # We also need to look for xhtml in the doctype line:
+        $regex = '/xhtml/i';
+        if (preg_match($regex, $line)) {
+            $not_html5 = 1;
+        }
+        fclose($tmpin);
+
+        # We got HTML5. Validate it:
+        if ($not_html5 == 0) {
+            print "<p style='color: green'>Validating your uploaded file...</p>";
+            append_output_log("Validating your uploaded file");
+            # TODO: what should we do if the validator fails? Log it? Error out?
+            system("$validator $basename >> $outfile 2>&1", $retval);
+            append_output_log("Validation complete");
+        }
+    }
+
+    if ($not_html5 || str_ends_with($basename, "xhtml")) {
+        echo <<<EOPARA
+        <p style='color: blue'>Your HTML does not seem to be HTML version
+        5. Be sure to run the appropriate validator, probably the W3C's
+        <a href="https://validator.w3.org">legacy HTML &amp; CSS
+        validator</a>.</p>
+        EOPARA;
+
+        append_output_log("--- HTML validation not done");
+        append_output_log("HTML validation not done since your source is not HTML5.");
+        append_output_log("You must use a validator such as https://validator.w3.org in addition to ebookmaker");
+        append_output_log("---");
+    }
+}
+
+function run_ebookmaker(
+    string $input_path,
+    string $output_dir,
+    string $logfile,
+    string $make_formats,
+    ?string $ebook,
+    ?string $title,
+    ?string $author
+): array
+{
+    global $prog;
+
+    $ebookmaker_log = "$output_dir/ebookmaker.log";
+
+    // Options to ebookmaker:
+    $opts = [
+        "--validate",
```

Comment:
@windymilla - I don't think Guiguts uses this option, should we make that optional via the API interface as well?

Reply:
I think GG should use that option if online ebookmaker does